### DISCUSSION TITLE
I want to use Copilot suggestion with <Tab> but it does not work properly

### DISCUSSION THREAD

**Discussion by aeciolevy** [LazyVim/LazyVim#4830] [Q&A] [answered]
Hi,

I am using LazyVim almost bare bones with plugins; however, I wanted to have copilot suggestions to be completed with `<Tab>` instead of going to the autocomplete panel and being mixed with variables and other things.

It works sometimes; to be honest, I cannot define what it makes work or not.

What I did to achieve that was disabling the `copilot-cmp`

```lua
-- disabled.lua file
return {
  -- add disabled plugins here
  { "zbirenbaum/copilot-cmp", enabled = false },
}
```

```lua
-- copilot.lua
return {
  {
    "zbirenbaum/copilot.lua",
    event = "InsertEnter",
    opts = {
      suggestion = {
        enabled = true,
        auto_trigger = true,
        keymap = {
          accept = "<Tab>",
          next = "<M-]>",
          prev = "<M-[>",
          accept_word = "<M-Tab>",
          dismiss = "<C-]>",
        },
      },
    },
    config = function(_, opts)
      -- Apply the opts as usual
      require("copilot").setup(opts)

      -- Check if the current `<Tab>` mapping is not Copilot's, then reset it
      local map = vim.api.nvim_get_keymap("i")
      local tab_mapped = false
      for _, m in pairs(map) do
        if m.lhs == "<Tab>" and m.desc == "Copilot Accept" then
          tab_mapped = true
          break
        end
      end
      if not tab_mapped then
        vim.keymap.set("i", "<Tab>", function()
          local suggestion = require("copilot.suggestion")
          if suggestion.is_visible() then
            suggestion.accept()
            return ""
          else
            return vim.api.nvim_replace_termcodes("<Tab>", true, true, true)
          end
        end, { expr = true, silent = true, desc = "Copilot Accept" })
      end
    end,
  },
}
```

This works sometimes; when it does not, I can see the `imap <Tab>` output this

```
i  <Tab>       * <Lua 199: ~/.local/share/nvim/lazy/lazy.nvim/lua/lazy/core/handler/keys.lua:121>
```

I would like a robust way to set it up. I have no idea what makes what I did the remap work or not. It depends on the project (some projects work, others don't), it depends on how I open Neovim. Totally random for my understanding.

I am not new to Vim/Neovim, but to customize Neovim, I mostly use StackOverflow and now ChatGPT.

Any help will be appreciated.

**Comment by folke** [comment, marked as answer]
Remove all your customizations and set vim.g.ai_cmp = false in your options.lua.

Dont forget to remove **all** the changes you made above, or it won't work

**Reply by aeciolevy** [reply]
I am gonna test right now

**Reply by aeciolevy** [reply]
I set it, but it is working back only with `<CR>` inside of the nvim-cmp.
I deleted all the customization I made with copilot.lua.  

![2024-11-18 11 58 34](https://github.com/user-attachments/assets/1d531938-e51a-40c6-9a16-a2230f30676a)

```lua
-- Options are automatically loaded before lazy.nvim startup
-- Default options that are always set: https://github.com/LazyVim/LazyVim/blob/main/lua/lazyvim/config/options.lua
-- Add any additional options here
vim.g.ai_cmp = false
```

**Reply by folke** [reply]
You need to delete all your customizations. All of them. To nvim CMP, copilot, and custom tab keymaps, snippet keymaps, ...

**Reply by folke** [reply]
And are you sure your plugins are up to date?

**Reply by aeciolevy** [reply]
I am gonna test and delete any other customization. I keep you posted here

**Reply by aeciolevy** [reply]
Thank you so much. It is working flawlessly.

**Reply by raRaRa** [reply]
@folke I'm coming from kickstart, tab has never worked out of the box for auto completion. Should I use the supertab recipe on the LazyVim website?
